### PR TITLE
380 - Refactor usePollDatasetStatus to accept a new parameter for polling check

### LINKED ISSUE #380: Refactor usePollDatasetStatus to accept a new parameter for polling check
### Context

When users start processing a dataset for download (i.e., one-off experiment, a shared dataset, My Dataset), the application displays the dataset processing view in the UI (refer to #175 and #200). To keep the UI updated, we use the `usePollDatasetId` hook to poll the latest dataset status.

Currently, the hook checks the dataset status using an accession code for one-off experiment download or the dataset ID by calling the `pollDatasetAccession`  and `pollDatasetId` methods from the components that use this hook.

e.g., For one-off downlods: 

```js
const { isProcessingDataset, pollDatasetAccession } = usePollDatasetStatus()
pollDatasetAccession(accessionCode) // checks this accession code for polling
```

### Problem or idea

To simplify this workflow, we should update the `usePollDatasetId` to accept a new parameter. This change will allow us to clean up method calls in the consuming components.

e.g., The new approach may look as follows:

```js
  const { isProcessingDataset } = usePollDatasetStatus(accessionCode) // pass an accession code here
  // eliminates the need for separate polling check
```

### Solution or next step

1. Update the `usePollDatasetId` hook to accept a new parameter, `accessionCodeOrDatasetId` 
2. Utilize `useEffect` within the hook to watch this newly added parameter and trigger polling as needed
3. Clean up the no longer required method calls from the following components:
   - `SearchCardAction`
   - `Dataset` (`pages/[dataset_id]`)

### PR DESCRIPTION
**Pull request by nozomione**
## Issue Number

Closes #380

## Purpose/Implementation Notes
I've added the new parameter `accessionCodeOrDatasetId` to `usePollDatasetStatus`  and adjusted the codeabse.

Changes include:
- Added the new `accessionCodeOrDatasetId` parameter to `usePollDatasetStatus` 
- Added a new `useEffect` block to initiate the dataset polling in `usePollDatasetStatus`
- Remove method calls from `[dataset_id]` and `SearchCardAction` components and adjusted the codebase

## Types of changes

- Refactor (addresses code organization and design mentioned in corresponding issue)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

### CODE REVIEW

File: src/hooks/usePollDatasetStatus.js [hunk 1]
```diff
@@ -47,8 +59,8 @@ export const usePollDatasetStatus = () => {
 
   const refreshProcessingDataset = async () => {
     const response = await getDataset(polledDatasetId)
-    // TEMP: until the fetchAsync is refactored
-    if (response?.ok !== false) {
```
Comment (nozomione) [inline src/hooks/usePollDatasetStatus.js:51]: I've cleaned up the no longer required optional chaining and the strict equality check here to align with the latest `dev` (merged PR #337) from `refreshProcessingDataset`, thank you David.

File: src/hooks/usePollDatasetStatus.js [hunk 2]
```diff
@@ -27,9 +33,15 @@ export const usePollDatasetStatus = () => {
     }
   }, [isProcessingDataset])
 
+  // sets the dataset ID via the given accessionCodeOrDatasetId
+  const pollDatasetStatus = () => {
+    pollDatasetAccession(accessionCodeOrDatasetId)
+    pollDatasetId(accessionCodeOrDatasetId)
+  }
+
```
Comment (davidsmejia) [inline src/hooks/usePollDatasetStatus.js:41]: lets remove this here

File: src/hooks/usePollDatasetStatus.js [hunk 3]
```diff
@@ -1,14 +1,20 @@
...
     useDatasetManager()
   const [polledDatasetId, setPolledDatasetId] = useState(null)
   const [polledDatasetState, setPolledDatasetState] = useState(null)
   const timerRef = useRef(null)
   const [isProcessingDataset, setIsProcessingDataset] = useState(false)
 
+  // starts polling the dataset status
+  useEffect(() => {
+    pollDatasetStatus()
+  }, [])
```
Comment (davidsmejia) [inline src/hooks/usePollDatasetStatus.js:16]:
```suggestion
  // starts polling the dataset status
  useEffect(() => {
    if (!accessionOrId) return
    if (accessionOrId in processingDatasets) pollDatasetId(accessionOrId)
    if (accessionOrId in datasetAccessions) pollDatasetAccession(accessionOrId)
  }, [accessionOrId])
```

File: src/hooks/usePollDatasetStatus.js [hunk 4]
```diff
@@ -1,14 +1,20 @@
 import { useEffect, useState, useRef } from 'react'
 import { useDatasetManager } from 'hooks/useDatasetManager'
 
-export const usePollDatasetStatus = () => {
+// takes either an accession code (one-off) or a dataset ID
+export const usePollDatasetStatus = (accessionCodeOrDatasetId) => {
```
Comment (davidsmejia) [inline src/hooks/usePollDatasetStatus.js:5]:
```suggestion
export const usePollDatasetStatus = (accessionOrId) => {
```

File: src/hooks/usePollDatasetStatus.js [hunk 5]
```diff
@@ -1,14 +1,20 @@
 import { useEffect, useState, useRef } from 'react'
 import { useDatasetManager } from 'hooks/useDatasetManager'
 
-export const usePollDatasetStatus = () => {
+// takes either an accession code (one-off) or a dataset ID
```
Comment (davidsmejia) [inline src/hooks/usePollDatasetStatus.js:4]:
```suggestion
// takes either an accession code or a dataset ID to poll / check status
```

File: src/components/shared/SearchCard/SearchCardAction/SearchCardAction.js [hunk 6]
```diff
@@ -14,9 +14,9 @@ export const SearchCardAction = ({ experiment, technology }) => {
     num_downloadable_samples: downloadableSamples,
     organism_names: organismNames
   } = experiment
-  const { isProcessingDataset, pollDatasetAccession } = usePollDatasetStatus()
-  pollDatasetAccession(accessionCode) // checks this accession code for polling
+  const { datasetAccessions } = usePollDatasetStatus(accessionCode)
   const { setResponsive } = useResponsive()
+  const isProcessingOneOff = accessionCode in datasetAccessions
```
Comment (davidsmejia) [inline src/components/shared/SearchCard/SearchCardAction/SearchCardAction.js:19]:
```suggestion
```

File: src/components/shared/SearchCard/SearchCardAction/SearchCardAction.js [hunk 7]
```diff
@@ -14,9 +14,9 @@ export const SearchCardAction = ({ experiment, technology }) => {
     num_downloadable_samples: downloadableSamples,
     organism_names: organismNames
   } = experiment
-  const { isProcessingDataset, pollDatasetAccession } = usePollDatasetStatus()
-  pollDatasetAccession(accessionCode) // checks this accession code for polling
+  const { datasetAccessions } = usePollDatasetStatus(accessionCode)
```
Comment (davidsmejia) [inline src/components/shared/SearchCard/SearchCardAction/SearchCardAction.js:17]:
```suggestion
  const { isProcessingDataset } = usePollDatasetStatus(accessionCode)
```

**Review by davidsmejia — COMMENTED** [review]
Now you can remove the guards inside of `pollDatasetId` because that should be programmatically correct when using those methods. If there is ambiguity it would be when we kick off the hook.

**Comment by nozomione** [comment]
I've applied your feedback and this PR is ready for another look. Thank you David!

**Review by davidsmejia — APPROVED** [review]
LGTM

**Pull request merged**